File: frontend2/types/sdgRanks.ts
export interface SDGRankSchemaBase {
  rankId: number;
  sdgGoalId: number;
  tier: number;
  name: string;
  description?: string | null;
  xpRequired: number;
}

export interface SDGRankSchemaFull extends SDGRankSchemaBase {
  createdAt: string;
  updatedAt: string;
}



// Not derived from model
export interface UserSDGRankSchemaFull {
  userId: number;
  ranks: SDGRankSchemaFull[]; // User-specific ranks
}



// REQUESTS
export interface SDGRankCreateRequest {
  sdgGoalId: number;  // ID of the SDG goal (1-17)
  tier: number;  // Rank tier (1, 2, 3)
  name: string;  // Rank name
  description?: string;  // Optional description
  xpRequired: number;  // XP required for this rank
}

File: schemas/sdg_ranks.py
from datetime import datetime
from pydantic import BaseModel
from typing import Optional, List, Union


class SDGRankSchemaBase(BaseModel):
    rank_id: int
    sdg_goal_id: int
    tier: int
    name: str
    description: Optional[str] = None
    xp_required: float

    model_config = {
        "from_attributes": True
    }


class SDGRankSchemaFull(SDGRankSchemaBase):
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True
    }

# Not directly derived derived from models
class UsersSDGRankSchemaBase(BaseModel):
    user_id: Optional[int]
    user: Optional[Union["UserSchemaBase", "UserSchemaFull"]]
    ranks: List[SDGRankSchemaFull]  # List of SDG ranks for that user

    model_config = {
        "from_attributes": True
    }
